File: src/hara/lang/data/Vector.java
package hara.lang.data;

import java.util.Iterator;

import java.util.NoSuchElementException;
import java.util.concurrent.atomic.AtomicReference;

import hara.lang.base.*;
import hara.lang.base.G.HashType;

public interface Vector<E> extends T.VectorType<E> {

	public interface Fn {

		// STATIC METHODS
		public static Node editableRoot(Node node) {
			return new Node(new AtomicReference<>(Thread.currentThread()), node.array.clone());
		}

		public static Node editableNode(Node node, AtomicReference<Thread> edit) {
			return (node.edit == edit) 
					? node
					: new Node(edit, node.array.clone());
		}

		public static int tailoff(int size) {
			return (size < Node.NODE_MAXLEN) 
					? 0 
					: ((size - 1) >>> Node.NODE_SHIFT) << Node.NODE_SHIFT;
		}
		
		@SuppressWarnings("unchecked")
		public static <E> E[] getNodeArrayFor(
				Node root, int size, int shift, 
				E[] tail, int i, boolean editable) {
			if (i >= 0 && i < size) {
				if (i >= Fn.tailoff(size)) {
					return tail;
				}
				Node node = root;
				for (int level = shift; level > 0; level -= Node.NODE_SHIFT) {
					node = (editable == false) 
							? (Node) node.array[(i >>> level) & Node.NODE_MASK]
							: editableNode((Node) node.array[(i >>> level) & Node.NODE_MASK], root.edit);
				}
				return (E[]) node.array;
			}
			throw new IndexOutOfBoundsException();
		}

		public static Node newPath(AtomicReference<Thread> edit, int level, Node node) {
			if (level == 0) return node;
			var ret = new Node(edit);
			ret.array[0] = newPath(edit, level - Node.NODE_SHIFT, node);
			return ret;
		}

		public static Object[] makeTail(Object[] tail) {
			var ret = new Object[32];
			System.arraycopy(tail, 0, ret, 0, tail.length);
			return ret;
		}
		
		public static Object[] newTail(Object[] tail, int i, Object val){
			var newTail = new Object[tail.length];
			System.arraycopy(tail, 0, newTail, 0, tail.length);
			newTail[i & Node.NODE_MASK] = val;
			return newTail;
		}
		
		public static Object[] newTailAppend(Object[] tail, Object val) {
			var newTail = new Object[tail.length + 1];
			System.arraycopy(tail, 0, newTail, 0, tail.length);
			newTail[tail.length] = val;
			return newTail;
		}

		
		public static Object[] trimTail(Object[] tail, int size) { 
			var trimmed = new Object[size - tailoff(size)];
			System.arraycopy(tail, 0, trimmed, 0, trimmed.length);
			return trimmed;
		}
		
		public static Node pushTail(
				Node root, int size, int level, 
				Node parent, Node _tailnode, boolean editable) {
			if (editable == true) parent = Fn.editableNode(parent, root.edit);
			int subidx = ((size - 1) >>> level) & Node.NODE_MASK;
			var ret = new Node(parent.edit, parent.array.clone());
			Node nodeToInsert;
			if (level == Node.NODE_SHIFT) {
				nodeToInsert = _tailnode;
			} else {
				Node child = (Node) parent.array[subidx];
				nodeToInsert = (child == null) 
						? Fn.newPath(root.edit, level - Node.NODE_SHIFT, _tailnode)
						: pushTail(root, size, level - Node.NODE_SHIFT, child, _tailnode, editable);
			}
			ret.array[subidx] = nodeToInsert;
			return ret;
		}

		public static Node popTail(
				Node root, int size, int level, 
				Node node, boolean editable) {
			if (editable == true) node = Fn.editableNode(node, root.edit);
			int subidx = ((size - 2) >>> level) & Node.NODE_MASK;
			if (level > Node.NODE_SHIFT) {
				Node newchild = popTail(root, size, level - Node.NODE_SHIFT, (Node) node.array[subidx], editable);
				if (newchild == null && subidx == 0)
					return null;
				else {
					node.array[subidx] = newchild;
					return node;
				}
			} else if (subidx == 0)
				return null;
			else {
				node.array[subidx] = null;
				return node;
			}
		}
		
		public static Node assoc(Node node, int level, int i, Object val) {
			var ret = new Node(node.edit, node.array.clone());
			if (level == 0) {
				ret.array[i & Node.NODE_MASK] = val;
			} else {
				int subidx = (i >>> level) & Node.NODE_MASK;
				ret.array[subidx] = assoc((Node) node.array[subidx], level - Node.NODE_SHIFT, i, val);
			}
			return ret;
		}
	}

	public class Node {
		public static final int NODE_SHIFT = 5;
		public static final int NODE_MAXLEN = 1 << NODE_SHIFT;
		public static final int NODE_MASK = NODE_MAXLEN - 1;

		public static final AtomicReference<Thread> NOEDIT = new AtomicReference<>(null);
		public static final Node EMPTY = new Node(NOEDIT, new Object[NODE_MAXLEN]);

		// DATA
		public transient final AtomicReference<Thread> edit;
		public final Object[] array;

		// CONSTRUCTORS
		public Node(AtomicReference<Thread> edit, Object[] array) {
			this.edit = edit;
			this.array = array;
		}

		public Node(AtomicReference<Thread> edit) {
			this.edit = edit;
			this.array = new Object[NODE_MAXLEN];
		}
	}

	// CONSTANTS
	public interface Base<E> extends Vector<E>, I.Assoc<Integer, E>, I.ObjType {

		public Node _root();

		public int _size();

		public int _shift();

		public E[] _tail();

		@Override
		default Base<E> conj(E e) {
			return (Base<E>) pushLast(e);
		}

		@Override
		default long count() {
			return _size();
		}
		
		default Iterator<E> rangedIterator(final int start, final int end) {
			return new Iterator<E>() {
				int i = start;
				int base = i - (i % 32);
				E[] array = (start < count()) 
						? Fn.getNodeArrayFor(_root(), _size(), _shift(), _tail(), i, false)
						: null;

				public boolean hasNext() {
					return i < end;
				}

				public E next() {
					if (i < end) {
						if (i - base == 32) {
							array = Fn.getNodeArrayFor(_root(), _size(), _shift(), _tail(), i, false);
							base += 32;
						}
						return array[i++ & 0x01f];
					} else {
						throw new NoSuchElementException();
					}
				}

				public void remove() {
					throw new UnsupportedOperationException();
				}
			};
		}

		@Override
		default Iterator<E> iterator() {
			return rangedIterator(0, _size());
		}
	}

	public class Mutable<E> extends T.RefType.MT implements Base<E>, I.ToPersistent {

		private int _size;
		private int _shift;
		private transient Node _root;
		private E[] _tail;

		@SuppressWarnings("unchecked")
		public Mutable(Base<E> v) {
			this(v.meta(), v._size(), v._shift(), Fn.editableRoot(v._root()), (E[]) Fn.makeTail(v._tail()));
		}

		public Mutable(I.Metadata meta, int size, int shift, Node root, E[] tail) {
			super(meta);
			_size = size;
			_shift = shift;
			_root = root;
			_tail = tail;
		}

		private void checkEditable() {
			if (_root.edit.get() == null) {
				throw new IllegalStateException("Mutable used after immutable! call");
			}
		}

		@Override
		public int _size() {
			return _size;
		}

		@Override
		public int _shift() {
			return _shift;
		}

		@Override
		public Node _root() {
			checkEditable();
			return _root;
		}

		@Override
		public E[] _tail() {
			checkEditable();
			return _tail;
		}

		@SuppressWarnings("unchecked")
		public Mutable<E> pushLast(E val) {
			checkEditable();
			int i = _size;
			// room in _tail?
			if (i - Fn.tailoff(_size) < Node.NODE_MAXLEN) {
				_tail[i & Node.NODE_MASK] = val;
				++_size;
				return this;
			}
			// full _tail, push into tree
			Node new_root;
			Node _tailnode = new Node(_root.edit, _tail);
			_tail = (E[]) new Object[Node.NODE_MAXLEN];
			_tail[0] = val;
			int new_shift = _shift;
			// overflow _root?
			if ((_size >>> Node.NODE_SHIFT) > (1 << _shift)) {
				new_root = new Node(_root.edit);
				new_root.array[0] = _root;
				new_root.array[1] = Fn.newPath(_root.edit, _shift, _tailnode);
				new_shift += Node.NODE_SHIFT;
			} else
				new_root = Fn.pushTail(_root, _size, _shift, _root, _tailnode, true);
			_root = new_root;
			_shift = new_shift;
			++_size;
			return this;
		}

		@Override
		public E nth(long i) {
			checkEditable();
			E[] node = Fn.getNodeArrayFor(_root, _size, _shift, _tail, (int) i, true);
			return node[(int) i & Node.NODE_MASK];
		}

		@Override
		public Mutable<E> assoc(Integer idx, E e) {
			checkEditable();
			E[] node = Fn.getNodeArrayFor(_root, _size, _shift, _tail, idx, true);
			node[idx & Node.NODE_MASK] = e;
			return this;
		}

		@Override
		public Mutable<E> popLast() {
			checkEditable();
			if (_size == 0)
				throw new IllegalStateException("Can't pop empty vector");
			if (_size == 1) {
				_size = 0;
				return this;
			}
			int i = _size - 1;
			if ((i & Node.NODE_MASK) > 0) {
				_size--;
				return this;
			}

			E[] newtail = Fn.getNodeArrayFor(_root, _size, _shift, _tail, _size - 2, true);

			Node newroot = Fn.popTail(_root, _size, _shift, _root, true);
			int newshift = _shift;
			if (newroot == null) {
				newroot = new Node(_root.edit);
			}
			if (_shift > Node.NODE_SHIFT && newroot.array[1] == null) {
				newroot = Fn.editableNode((Node) newroot.array[0], _root.edit);
				newshift -= Node.NODE_SHIFT;
			}
			_root = newroot;
			_shift = newshift;
			_size--;
			_tail = newtail;
			return this;
		}

		@SuppressWarnings("unchecked")
		@Override
		public Standard<E> toPersistent() {
			_root.edit.set(null);
			return new Standard<>(_meta, _size, _shift, _root, (E[]) Fn.trimTail((Object[])_tail, _size));
		}

		@SuppressWarnings("unchecked")
		@Override
		public Mutable<E> empty() {
			return new Mutable<E>(null, 0, Node.NODE_SHIFT, Node.EMPTY, (E[]) new Object[] {});
		}
	}

	public class Standard<E> extends T.RefType.PT implements Base<E>, I.ToMutable {

		// STATIC
		public final static Standard<Object> EMPTY = new Standard<>(null, 0, Node.NODE_SHIFT, Node.EMPTY, new Object[] {});

		// INSTANCE
		private final int _size;
		private final int _shift;
		private final transient Node _root;
		private final E[] _tail;

		public Standard(Base<E> v) {
			this(v.meta(), v._size(), v._shift(), v._root(), v._tail());
		}

		protected Standard(I.Metadata meta, int size, int shift, Node root, E[] tail) {
			super(meta);
			_size = size;
			_shift = shift;
			_root = root;
			_tail = tail;
		}

		@SuppressWarnings("unchecked")
		public static <E> Standard<E> empty(I.Metadata meta) {
			Standard<E> ret = (Standard<E>) EMPTY;
			return (meta == null) ? ret : ret.withMeta(meta);
		}

		@Override
		public int _size() {
			return _size;
		}

		@Override
		public int _shift() {
			return _shift;
		}

		@Override
		public Node _root() {
			return _root;
		}

		@Override
		public E[] _tail() {
			return _tail;
		}

		@Override
		public Standard<E> withMeta(I.Metadata meta) {
			return (meta() == meta) ? this : new Standard<E>(meta, _size, _shift, _root, _tail);
		}

		@Override
		public E nth(long i) {
			E[] node = Fn.getNodeArrayFor(_root, _size, _shift, _tail, (int) i, false);
			return node[(int) i & Node.NODE_MASK];
		}

		@SuppressWarnings("unchecked")
		@Override
		public Standard<E> assoc(Integer i, E val) {
			if (i >= 0 && i < _size) {
				return (i >= Fn.tailoff(_size))
					? new Standard<>(_meta, _size, _shift, _root, (E[]) Fn.newTail(_tail, i, val))
					: new Standard<>(_meta, _size, _shift, Fn.assoc(_root, _shift, i, val), _tail);
			}
			if (i == _size) {
				return pushLast(val);
			}
			throw new IndexOutOfBoundsException();
		}

		@SuppressWarnings("unchecked")
		@Override
		public Standard<E> pushLast(E val) {
			if (_size - Fn.tailoff(_size) < Node.NODE_MAXLEN) {
				return new Standard<>(_meta, _size + 1, _shift, _root, 
										(E[]) Fn.newTailAppend((Object[])_tail, val));
			}

			Node new_root;
			Node _tailnode = new Node(_root.edit, _tail);
			int new_shift = _shift;
			if ((_size >>> Node.NODE_SHIFT) > (1 << _shift)) {
				new_root = new Node(_root.edit);
				new_root.array[0] = _root;
				new_root.array[1] = Fn.newPath(_root.edit, _shift, _tailnode);
				new_shift += Node.NODE_SHIFT;
			} else {
				new_root = Fn.pushTail(_root, _size, _shift, _root, _tailnode, false);
			}
			return new Standard<>(_meta, _size + 1, new_shift, new_root, (E[]) new Object[] { val });
		}

		@SuppressWarnings("unchecked")
		@Override
		public Standard<E> popLast() {
			if (_size == 0)
				throw new IllegalStateException("Can't pop empty vector");
			if (_size == 1)
				return empty(_meta);
			if (_size - Fn.tailoff(_size) > 1) {
				E[] newTail = (E[]) new Object[_tail.length - 1];
				System.arraycopy(_tail, 0, newTail, 0, newTail.length);
				return new Standard<E>(_meta, _size - 1, _shift, _root, newTail);
			}
			E[] newtail = Fn.getNodeArrayFor(_root, _size, _shift, _tail, _size - 2, false);

			Node new_root = Fn.popTail(_root, _size, _shift, _root, false);
			int new_shift = _shift;
			if (new_root == null) {
				new_root = Node.EMPTY;
			}
			if (_shift > Node.NODE_SHIFT && new_root.array[1] == null) {
				new_root = (Node) new_root.array[0];
				new_shift -= Node.NODE_SHIFT;
			}
			return new Standard<E>(_meta, _size - 1, new_shift, new_root, newtail);
		}

		@Override
		public Standard<E> empty() {
			return empty(_meta);
		}

		@Override
		public Mutable<E> toMutable() {
			return new Mutable<E>(this);
		}
	}

}

File: wasm/src/lang/data/vector.rs
use std::cell::Cell;
use std::rc::Rc;

use crate::lang::protocol::hash::HashType;
use crate::lang::protocol::{
    IAssoc, IConj, ICount, IDisplay, IEmpty, IEquality, IHash, IMetadata, IMutable, INth,
    IPeekFirst, IPeekLast, IPersistent, IPopLast, IPushLast, IToMutable, IToPersistent,
};

const NODE_SHIFT: usize = 5;
const NODE_WIDTH: usize = 1 << NODE_SHIFT;
const NODE_MASK: usize = NODE_WIDTH - 1;

#[derive(Debug, Clone)]
enum Node<E> {
    Branch(Rc<Vec<Option<Rc<Node<E>>>>>),
    Leaf(Rc<Vec<E>>),
}

impl<E> Node<E> {
    fn empty() -> Rc<Self> {
        Rc::new(Self::Branch(Rc::new(vec![None; NODE_WIDTH])))
    }
}

#[derive(Debug, Clone)]
pub struct Standard<E> {
    metadata: Option<Rc<crate::lang::data::Metadata>>,
    size: usize,
    shift: usize,
    root: Rc<Node<E>>,
    tail: Rc<Vec<E>>,
}

impl<E: Clone> Default for Standard<E> {
    fn default() -> Self {
        Self {
            metadata: None,
            size: 0,
            shift: NODE_SHIFT,
            root: Node::empty(),
            tail: Rc::new(Vec::new()),
        }
    }
}

impl<E: Clone> Standard<E> {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn from_iter(values: impl IntoIterator<Item = E>) -> Self {
        values
            .into_iter()
            .fold(Self::new(), |vector, value| vector.push_last(value))
    }

    pub fn len(&self) -> usize {
        self.size
    }

    pub fn is_empty(&self) -> bool {
        self.size == 0
    }

    fn tail_offset(&self) -> usize {
        if self.size < NODE_WIDTH {
            0
        } else {
            ((self.size - 1) >> NODE_SHIFT) << NODE_SHIFT
        }
    }

    pub fn get(&self, index: usize) -> Option<&E> {
        if index >= self.size {
            return None;
        }
        if index >= self.tail_offset() {
            return self.tail.get(index & NODE_MASK);
        }

        let mut node = self.root.as_ref();
        let mut level = self.shift;
        while level > 0 {
            let Node::Branch(children) = node else {
                return None;
            };
            node = children[(index >> level) & NODE_MASK].as_deref()?;
            level -= NODE_SHIFT;
        }
        match node {
            Node::Leaf(values) => values.get(index & NODE_MASK),
            Node::Branch(_) => None,
        }
    }

    pub fn assoc_value(&self, index: usize, value: E) -> Option<Self> {
        if index == self.size {
            return Some(self.push_last(value));
        }
        if index >= self.size {
            return None;
        }
        if index >= self.tail_offset() {
            let mut tail = (*self.tail).clone();
            tail[index & NODE_MASK] = value;
            return Some(Self {
                tail: Rc::new(tail),
                ..self.clone()
            });
        }
        Some(Self {
            root: assoc_node(&self.root, self.shift, index, value),
            ..self.clone()
        })
    }

    pub fn push_last(&self, value: E) -> Self {
        if self.tail.len() < NODE_WIDTH {
            let mut tail = (*self.tail).clone();
            tail.push(value);
            return Self {
                size: self.size + 1,
                tail: Rc::new(tail),
                ..self.clone()
            };
        }

        let tail_node = Rc::new(Node::Leaf(self.tail.clone()));
        let overflow = (self.size >> NODE_SHIFT) > (1usize << self.shift);
        let (root, shift) = if overflow {
            let mut children = vec![None; NODE_WIDTH];
            children[0] = Some(self.root.clone());
            children[1] = Some(new_path(self.shift, tail_node));
            (
                Rc::new(Node::Branch(Rc::new(children))),
                self.shift + NODE_SHIFT,
            )
        } else {
            (
                push_tail(&self.root, self.shift, self.size, tail_node),
                self.shift,
            )
        };

        Self {
            metadata: self.metadata.clone(),
            size: self.size + 1,
            shift,
            root,
            tail: Rc::new(vec![value]),
        }
    }

    pub fn pop_last_value(&self) -> Option<Self> {
        if self.size == 0 {
            return None;
        }
        if self.size == 1 {
            return Some(Self {
                metadata: self.metadata.clone(),
                ..Self::new()
            });
        }
        if self.size - self.tail_offset() > 1 {
            let mut tail = (*self.tail).clone();
            tail.pop();
            return Some(Self {
                size: self.size - 1,
                tail: Rc::new(tail),
                ..self.clone()
            });
        }

        let new_tail = self
            .array_for(self.size - 2)
            .expect("previous vector leaf")
            .clone();
        let mut root = pop_tail(&self.root, self.shift, self.size).unwrap_or_else(Node::empty);
        let mut shift = self.shift;
        if shift > NODE_SHIFT {
            if let Node::Branch(children) = root.as_ref() {
                if children[1].is_none() {
                    root = children[0].clone().expect("collapsed vector root");
                    shift -= NODE_SHIFT;
                }
            }
        }
        Some(Self {
            metadata: self.metadata.clone(),
            size: self.size - 1,
            shift,
            root,
            tail: Rc::new(new_tail),
        })
    }

    fn array_for(&self, index: usize) -> Option<&Vec<E>> {
        if index >= self.size {
            return None;
        }
        if index >= self.tail_offset() {
            return Some(&self.tail);
        }
        let mut node = self.root.as_ref();
        let mut level = self.shift;
        while level > 0 {
            let Node::Branch(children) = node else {
                return None;
            };
            node = children[(index >> level) & NODE_MASK].as_deref()?;
            level -= NODE_SHIFT;
        }
        match node {
            Node::Leaf(values) => Some(values),
            Node::Branch(_) => None,
        }
    }

    pub fn iter(&self) -> Iter<'_, E> {
        Iter {
            vector: self,
            index: 0,
        }
    }

    pub fn subview(&self, start: usize, end: usize) -> Option<SubView<E>> {
        if start > end || end > self.size {
            return None;
        }
        Some(SubView {
            vector: self.clone(),
            start,
            end,
        })
    }

    #[cfg(test)]
    fn shares_root_with(&self, other: &Self) -> bool {
        Rc::ptr_eq(&self.root, &other.root)
    }
}

fn new_path<E: Clone>(level: usize, node: Rc<Node<E>>) -> Rc<Node<E>> {
    if level == 0 {
        return node;
    }
    let mut children = vec![None; NODE_WIDTH];
    children[0] = Some(new_path(level - NODE_SHIFT, node));
    Rc::new(Node::Branch(Rc::new(children)))
}

fn push_tail<E: Clone>(
    parent: &Rc<Node<E>>,
    level: usize,
    size: usize,
    tail: Rc<Node<E>>,
) -> Rc<Node<E>> {
    let Node::Branch(existing) = parent.as_ref() else {
        unreachable!("vector tree parent must be a branch")
    };
    let mut children = (**existing).clone();
    let index = ((size - 1) >> level) & NODE_MASK;
    children[index] = Some(if level == NODE_SHIFT {
        tail
    } else if let Some(child) = &children[index] {
        push_tail(child, level - NODE_SHIFT, size, tail)
    } else {
        new_path(level - NODE_SHIFT, tail)
    });
    Rc::new(Node::Branch(Rc::new(children)))
}

fn assoc_node<E: Clone>(node: &Rc<Node<E>>, level: usize, index: usize, value: E) -> Rc<Node<E>> {
    if level == 0 {
        let Node::Leaf(existing) = node.as_ref() else {
            unreachable!("vector terminal node must be a leaf")
        };
        let mut values = (**existing).clone();
        values[index & NODE_MASK] = value;
        return Rc::new(Node::Leaf(Rc::new(values)));
    }
    let Node::Branch(existing) = node.as_ref() else {
        unreachable!("vector path must contain branches")
    };
    let mut children = (**existing).clone();
    let child_index = (index >> level) & NODE_MASK;
    children[child_index] = Some(assoc_node(
        children[child_index]
            .as_ref()
            .expect("existing vector path"),
        level - NODE_SHIFT,
        index,
        value,
    ));
    Rc::new(Node::Branch(Rc::new(children)))
}

fn pop_tail<E: Clone>(node: &Rc<Node<E>>, level: usize, size: usize) -> Option<Rc<Node<E>>> {
    let Node::Branch(existing) = node.as_ref() else {
        unreachable!("vector path must contain branches")
    };
    let index = ((size - 2) >> level) & NODE_MASK;
    if level > NODE_SHIFT {
        let child = pop_tail(
            existing[index].as_ref().expect("existing vector path"),
            level - NODE_SHIFT,
            size,
        );
        if child.is_none() && index == 0 {
            return None;
        }
        let mut children = (**existing).clone();
        children[index] = child;
        Some(Rc::new(Node::Branch(Rc::new(children))))
    } else if index == 0 {
        None
    } else {
        let mut children = (**existing).clone();
        children[index] = None;
        Some(Rc::new(Node::Branch(Rc::new(children))))
    }
}

pub struct Iter<'a, E> {
    vector: &'a Standard<E>,
    index: usize,
}

impl<'a, E: Clone> Iterator for Iter<'a, E> {
    type Item = &'a E;

    fn next(&mut self) -> Option<Self::Item> {
        let value = self.vector.get(self.index);
        self.index += usize::from(value.is_some());
        value
    }
}

impl<E: Clone> FromIterator<E> for Standard<E> {
    fn from_iter<T: IntoIterator<Item = E>>(iter: T) -> Self {
        Self::from_iter(iter)
    }
}

impl<E: Clone> From<Vec<E>> for Standard<E> {
    fn from(values: Vec<E>) -> Self {
        Self::from_iter(values)
    }
}

impl<E: Clone> std::ops::Index<usize> for Standard<E> {
    type Output = E;

    fn index(&self, index: usize) -> &Self::Output {
        self.get(index).expect("vector index out of bounds")
    }
}

impl<E: Clone + PartialEq> PartialEq for Standard<E> {
    fn eq(&self, other: &Self) -> bool {
        self.size == other.size && self.iter().eq(other.iter())
    }
}

impl<E: Clone + PartialEq> IEquality for Standard<E> {
    fn equality(&self, other: &Self) -> bool {
        self == other
    }
}

impl<E: Clone> ICount for Standard<E> {
    fn count(&self) -> usize {
        self.size
    }
}

impl<E: Clone> INth<E> for Standard<E> {
    fn nth(&self, index: usize) -> Option<&E> {
        self.get(index)
    }
}

impl<E: Clone> IAssoc<usize, E> for Standard<E> {
    type Output = Self;
    fn assoc(&self, index: usize, value: E) -> Self {
        self.assoc_value(index, value)
            .expect("vector index out of bounds")
    }
}

impl<E: Clone> IPeekFirst<E> for Standard<E> {
    fn peek_first(&self) -> Option<E> {
        self.get(0).cloned()
    }
}
impl<E: Clone> IPeekLast<E> for Standard<E> {
    fn peek_last(&self) -> Option<E> {
        self.size
            .checked_sub(1)
            .and_then(|index| self.get(index))
            .cloned()
    }
}
impl<E: Clone> IPushLast<E> for Standard<E> {
    type Output = Self;
    fn push_last(&self, value: E) -> Self {
        Standard::push_last(self, value)
    }
}

impl<E: Clone> IPopLast for Standard<E> {
    type Output = Self;
    fn pop_last(&self) -> Self {
        self.pop_last_value().expect("cannot pop empty vector")
    }
}

impl<E: Clone> IConj<E> for Standard<E> {
    type Output = Self;
    fn conj(&self, value: E) -> Self {
        self.push_last(value)
    }
}

impl<E: Clone> IEmpty for Standard<E> {
    type Output = Self;
    fn empty(&self) -> Self {
        Self {
            metadata: self.metadata.clone(),
            ..Self::new()
        }
    }
}

impl<E: Clone> IMetadata for Standard<E> {
    type Metadata = Rc<crate::lang::data::Metadata>;

    fn meta(&self) -> Option<&Self::Metadata> {
        self.metadata.as_ref()
    }

    fn with_meta(&self, metadata: Option<Self::Metadata>) -> Self {
        Self {
            metadata,
            ..self.clone()
        }
    }
}

impl<E: Clone> IPersistent for Standard<E> {}

impl<E: Clone> IToMutable for Standard<E> {
    type Mutable = Mutable<E>;

    fn to_mutable(&self) -> Self::Mutable {
        Mutable::from_standard(self)
    }
}

impl<E: Clone + std::fmt::Debug> IDisplay for Standard<E> {
    fn display(&self) -> String {
        format!(
            "[{}]",
            self.iter()
                .map(|value| format!("{value:?}"))
                .collect::<Vec<_>>()
                .join(" ")
        )
    }
}

impl<E: Clone + std::hash::Hash> IHash for Standard<E> {
    fn hash_calc(&self, _hash_type: HashType) -> u64 {
        use std::hash::{Hash, Hasher};
        let mut state = std::collections::hash_map::DefaultHasher::new();
        self.iter().for_each(|value| value.hash(&mut state));
        state.finish()
    }
}

#[derive(Debug, Clone)]
pub struct Mutable<E> {
    editable: Rc<Cell<bool>>,
    values: Vec<E>,
    metadata: Option<Rc<crate::lang::data::Metadata>>,
}

impl<E: Clone> Mutable<E> {
    fn from_standard(vector: &Standard<E>) -> Self {
        Self {
            editable: Rc::new(Cell::new(true)),
            values: vector.iter().cloned().collect(),
            metadata: vector.metadata.clone(),
        }
    }

    fn check_editable(&self) {
        assert!(
            self.editable.get(),
            "mutable vector used after to_persistent"
        );
    }

    pub fn push_last(&mut self, value: E) -> &mut Self {
        self.check_editable();
        self.values.push(value);
        self
    }

    pub fn assoc(&mut self, index: usize, value: E) -> &mut Self {
        self.check_editable();
        if index == self.values.len() {
            return self.push_last(value);
        }
        self.values[index] = value;
        self
    }
}

impl<E: Clone> IMutable for Mutable<E> {}

impl<E: Clone> IToPersistent for Mutable<E> {
    type Persistent = Standard<E>;

    fn to_persistent(&mut self) -> Self::Persistent {
        self.check_editable();
        self.editable.set(false);
        let mut vector = Standard::from_iter(self.values.clone());
        vector.metadata = self.metadata.clone();
        vector
    }
}

#[derive(Debug, Clone)]
pub struct SubView<E> {
    vector: Standard<E>,
    start: usize,
    end: usize,
}

impl<E: Clone> SubView<E> {
    pub fn len(&self) -> usize {
        self.end - self.start
    }

    pub fn get(&self, index: usize) -> Option<&E> {
        if index >= self.len() {
            None
        } else {
            self.vector.get(self.start + index)
        }
    }
}

#[cfg(test)]
mod tests {
    use super::Standard;
    use crate::lang::protocol::{IAssoc, IToMutable, IToPersistent};

    #[test]
    fn preserves_values_across_java_tree_boundaries() {
        let vector = (0..1057).collect::<Standard<_>>();
        let appended = vector.push_last(1057);
        let updated = appended.assoc(32, -1);

        assert_eq!(vector.len(), 1057);
        assert_eq!(vector[32], 32);
        assert_eq!(appended[1057], 1057);
        assert_eq!(updated[32], -1);
        assert_eq!(updated[1057], 1057);
    }

    #[test]
    fn tail_updates_share_the_tree() {
        let vector = (0..33).collect::<Standard<_>>();
        let appended = vector.push_last(33);

        assert!(vector.shares_root_with(&appended));
    }

    #[test]
    #[should_panic(expected = "mutable vector used after to_persistent")]
    fn mutable_vector_is_invalid_after_persisting() {
        let vector = (0..4).collect::<Standard<_>>();
        let mut mutable = vector.to_mutable();
        let _persistent = mutable.to_persistent();
        mutable.push_last(5);
    }
}
